### PR TITLE
Clamping hidden state values to allow FP16

### PR DESCRIPTION
**Pull request by SSamDav**
# What does this PR do?

Fixes # (issue)
Following the discussion in [#9295](https://github.com/huggingface/transformers/issues/9295) and the solution proposed by  [#9487](https://github.com/huggingface/transformers/pull/9487). 

Implement the solution that enables the FP16 for LongT5 models.

Who can review:
@patrickvonplaten, @patil-suraj

### CODE REVIEW

**Comment by SSamDav** [comment]
Hi, it seems that I have a test error, however, I didn't change the code that is falling. 
Does anyone know how I can pass these tests?

File: src/transformers/models/longt5/modeling_longt5.py [hunk 1]
```diff
@@ -1199,6 +1199,11 @@ def forward(
         hidden_states, present_key_value_state = self_attention_outputs[:2]
         attention_outputs = self_attention_outputs[2:]  # Keep self-attention outputs and relative position weights
 
+        # clamp inf values to enable fp16 training
+        if torch.isinf(hidden_states).any():
```
Comment (ydshieh) [inline src/transformers/models/longt5/modeling_longt5.py:1203]: Is there any reason not to include `if hidden_states.dtype == torch.float16` in the condition, which is done for `T5`

https://github.com/huggingface/transformers/blob/e1dd784b5c6f2c356835f5200d297e08058eb18c/src/transformers/models/t5/modeling_t5.py#L686

Same question for the other 2 places.

Comment (SSamDav) [reply src/transformers/models/longt5/modeling_longt5.py:1203]: No, probably I saw an old version of the code. 😅 
Will change the code to include the `torch.float16`

**Comment by ydshieh** [comment]
Thanks @SSamDav.

And sorry, I forgot to mention in order to avoid the test failure you saw yesterday, you can rebase the working branch on `main`.

**Review by ydshieh — APPROVED** [review]
Thank you @SSamDav 🚀 .

The fix follows exactly what has been done for `T5` 👍

**Comment by patrickvonplaten** [comment]
Ok for me if we know that it helps. In T5 it didn't really help.

Also cc @ArthurZucker here just FYI

**Comment by ydshieh** [comment]
Since @patrickvonplaten prefers to have an issue where this PR will solve, I think we are not going to merge this PR at this moment. Let's see if there will be such issues reported for `LongT5` in the future. We can make an investigation and decide if to re-open/merge this PR (or with a different fix). WDYT? cc @younesbelkada @ArthurZucker .

----

And regarding `nan`: My though is that it's most likely coming from the sequences with all `-inf` after adding the mask to the attention scores, and `nan` after `softmax`, like what we observed in OPT or  Bloom recently, where we provided a fix `as close as to` where `nan` happens. (However, in the case of `T5`, a clamp is done `T5LayerFF` which is not attention-related).

**Comment by SSamDav** [comment]
I my tests when I run a finetuned version of the `google/long-t5-tglobal-base` in FP16 I got `Nan`  in the forward step, I could check if the values com from the `LongT5LayerFF`.

**Comment by ydshieh** [comment]
> I my tests when I run a finetuned version of the `google/long-t5-tglobal-base` in FP16 I got `Nan` in the forward step, I could check if the values com from the `LongT5LayerFF`.

Hi, so it is running the inference, right? Is that finetuned checkpoint uploaded to Hub?

**Comment by SSamDav** [comment]
> Hi, so it is running the inference, right? 
Yes
>  Is that finetuned checkpoint uploaded to Hub?

**Comment by ydshieh** [comment]
> > Is that finetuned checkpoint uploaded to Hub?
> 
> No, it was trained in confidential data.

Got it. However, it would be really nice if we have a public available checkpoint (on another dataset) that can show the issue and the effect of the fix. I understand that it may not easy to obtain another such checkpoint - and potentially time consuming.

@patrickvonplaten Any further comment?

**Comment by younesbelkada** [comment]
Hi!
I second what @ydshieh said, probably the root cause of this is happening inside the attention score computation as it has been observed for BLOOM and OPT - maybe it's worth investigating a bit before merging! 
As a simple test, we could try to reproduce what has been done in #18057 & #17437 and see if this fixes the initial issue

File: src/transformers/models/longt5/modeling_longt5.py [hunk 2]
```diff
@@ -1199,6 +1199,11 @@ def forward(
         hidden_states, present_key_value_state = self_attention_outputs[:2]
         attention_outputs = self_attention_outputs[2:]  # Keep self-attention outputs and relative position weights
 
+        # clamp inf values to enable fp16 training
```
Comment (younesbelkada) [inline src/transformers/models/longt5/modeling_longt5.py:1202]:
```suggestion
        # clamp inf values to enable fp16 inference - check https://github.com/huggingface/transformers/pull/19229/
```
Maybe let's specify that it's just for inference?

File: src/transformers/models/longt5/modeling_longt5.py [hunk 3]
```diff
@@ -1221,6 +1226,11 @@ def forward(
             )
             hidden_states = cross_attention_outputs[0]
 
+            # clamp inf values to enable fp16 training
```
Comment (younesbelkada) [inline src/transformers/models/longt5/modeling_longt5.py:1229]:
```suggestion
            # clamp inf values to enable fp16 inference - check https://github.com/huggingface/transformers/pull/19229/
```
Same as above! ;)

File: src/transformers/models/longt5/modeling_longt5.py [hunk 4]
```diff
@@ -1231,6 +1241,11 @@ def forward(
         # Apply Feed Forward layer
         hidden_states = self.layer[-1](hidden_states)
 
+        # clamp inf values to enable fp16 training
```
Comment (younesbelkada) [inline src/transformers/models/longt5/modeling_longt5.py:1244]:
```suggestion
       # clamp inf values to enable fp16 inference - check https://github.com/huggingface/transformers/pull/19229/
```

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @SSamDav for the fix! 
I managed to reproduce the initial issue with the following snippet:
```
import torch
from transformers import AutoTokenizer, LongT5ForConditionalGeneration

tokenizer = AutoTokenizer.from_pretrained("google/long-t5-tglobal-base")

model = LongT5ForConditionalGeneration.from_pretrained("google/long-t5-tglobal-base", torch_dtype=torch.float16).to(0)
inputs = tokenizer(100 * "studies have shown that owning a dog is good for you ", return_tensors="pt")
input_ids = inputs.input_ids

outputs = model.encoder(input_ids.to(0))
print(outputs.last_hidden_state.isnan().any())
```
However, it seems that the rootcause of this issue is happening at the `LongT5LayerFF` layer, exactly at [this line](https://github.com/huggingface/transformers/blob/5cd16f01db3b5499d4665e8624801ed30ba87bdd/src/transformers/models/longt5/modeling_longt5.py#L318). It seems that adding the previous hidden states with the GeLU-ed hidden states (`forwarded_states`) causes overflow issues. I tried adding `scores = torch.max(scores, torch.tensor(torch.finfo(scores.dtype).min))` [here](https://github.com/huggingface/transformers/blob/5cd16f01db3b5499d4665e8624801ed30ba87bdd/src/transformers/models/longt5/modeling_longt5.py#L719) but this seems to not help as the overflow comes after the attention layer. I propose to use these changes for now as it definitely helps to get inference in fp16 working. Maybe we should still add the line `scores = torch.max(scores, torch.tensor(torch.finfo(scores.dtype).min))`, but I can see that the attention scores are casted in fp32 before the softmax, so maybe it's not necessary - cc @ydshieh 🙏 

I think that we should also add a small slow test reproducing the behavior of the snippet above!
```
@slow 
def test_fp16_inference(self):
    tokenizer = AutoTokenizer.from_pretrained("google/long-t5-tglobal-base")
    model = LongT5ForConditionalGeneration.from_pretrained("google/long-t5-tglobal-base", torch_dtype=torch.float16).to(0)
    inputs = tokenizer(100 * "studies have shown that owning a dog is good for you ", return_tensors="pt")
    input_ids = inputs.input_ids
    outputs = model.encoder(input_ids.to(0))
    self.assertFalse(outputs.last_hidden_state.isnan().any())
```

I also propose to change the comments and explicitly specify that this helps for fp16 inference, not training as mentioned by @patrickvonplaten

**Comment by patrickvonplaten** [comment]
Hey @SSamDav,

If the PR as is now solves your problem for **inference** - it's good for me to merge! I don't think it'll fix problems with fine-tuning though

**Comment by SSamDav** [comment]
Hey @patrickvonplaten, good thanks for the help!
